README.md
<h1 align="center" style="border-bottom: none;">📦 topgun-buffer</h1>
<h3 align="center">An isomorphic Buffer API for <a href="https://github.com/TopGunBuild/topgun">TopGun</a></h3>

<p align="center">
  <a href="https://github.com/semantic-release/semantic-release">
      <img alt="semantic-release" src="https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg">
  </a>
  <a href="https://npm.im/topgun-buffer">
    <img alt="npm" src="https://badgen.net/npm/v/topgun-buffer">
  </a>
  <a href="https://bundlephobia.com/result?p=topgun-buffer">
    <img alt="bundlephobia" src="https://img.shields.io/bundlephobia/minzip/topgun-buffer.svg">
  </a>
  <a href="https://opensource.org/licenses/MIT">
      <img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-yellow.svg">
  </a>
</p>

This is Buffer implementation used in <a href="https://github.com/TopGunBuild/topgun">TopGun</a>. Functionality is mostly compatible with NodeJS 'safe-buffer' and is used for encoding conversions between binary and 'hex' | 'utf8' | 'base64'.

## Install

`npm install topgun-buffer`

## License

MIT


CHANGELOG.md
## [1.0.6](https://github.com/TopGunBuild/topgun-buffer/compare/v1.0.5...v1.0.6) (2023-05-03)


### Bug Fixes

* update typings ([d1b2ffd](https://github.com/TopGunBuild/topgun-buffer/commit/d1b2ffd2239c4fe9cd8c0bc2c657c8f03c95f8e3))
* update typings ([78b8115](https://github.com/TopGunBuild/topgun-buffer/commit/78b8115446f8b39cf79bde978ef10582a0de540d))

## [1.0.5](https://github.com/TopGunBuild/topgun-buffer/compare/v1.0.4...v1.0.5) (2023-05-03)


### Bug Fixes

* publish typings ([cbfef23](https://github.com/TopGunBuild/topgun-buffer/commit/cbfef232e7332cacad6fe1971f84bc462aa063cc))

## [1.0.4](https://github.com/TopGunBuild/topgun-buffer/compare/v1.0.3...v1.0.4) (2023-05-03)


### Bug Fixes

* rewrite buffer implementation ([530c9e4](https://github.com/TopGunBuild/topgun-buffer/commit/530c9e47e249add34f431953bf9aef52ff5eb755))

## [1.0.3](https://github.com/TopGunBuild/topgun-buffer/compare/v1.0.2...v1.0.3) (2023-04-18)


### Bug Fixes

* fix typings ([bb00605](https://github.com/TopGunBuild/topgun-buffer/commit/bb0060518cffbd53d860453175fbf7c0cb33016c))
* globalName "topGunBuffer" -> "TopGunBuffer" ([7e2628b](https://github.com/TopGunBuild/topgun-buffer/commit/7e2628bcf56c689af269a45522c6e7b7c5fa48bf))
* update tsup config for types generation ([be10a53](https://github.com/TopGunBuild/topgun-buffer/commit/be10a53ad8953cdf8828586064f834a48b105c81))

## [1.0.2](https://github.com/TopGunBuild/topgun-buffer/compare/v1.0.1...v1.0.2) (2023-04-18)


### Bug Fixes

* update "exports" in package.json ([94e9c77](https://github.com/TopGunBuild/topgun-buffer/commit/94e9c7771bc1b91ce0941e3857046c2e7910b098))

## [1.0.1](https://github.com/TopGunBuild/topgun-buffer/compare/v1.0.0...v1.0.1) (2023-04-17)


### Bug Fixes

* update config in package.json ([4519e22](https://github.com/TopGunBuild/topgun-buffer/commit/4519e22cf98d78c939429488fbbfffe179c0f967))

# 1.0.0 (2023-04-17)


### Features

* add Buffer class ([678a34f](https://github.com/TopGunBuild/topgun-buffer/commit/678a34f554fccb232228c1400c8f693859dc3f14))

# 1.0.0 (2023-04-09)


### Features

* initial ([95d7b1a](https://github.com/TopGunBuild/topgun-webcrypto/commit/95d7b1a1a1b4417b01fc0f1121675dbd885aa4fa))


test/index.test.ts
import Buffer from '../src/browser';

describe('Buffer', () => {
    it('indexes from a string', async () =>
    {
        const buf = Buffer.from('abc');
        expect(buf[0]).toEqual(97);
        expect(buf[1]).toEqual(98);
        expect(buf[2]).toEqual(99);
    });

    it('indexes from an array', async () =>
    {
        const buf = Buffer.from([97, 98, 99]);
        expect(buf[0]).toEqual(97);
        expect(buf[1]).toEqual(98);
        expect(buf[2]).toEqual(99);
    });

    it('setting index value should modify buffer contents', async () =>
    {
        const buf = Buffer.from([97, 98, 99]);
        expect(buf[2]).toEqual(99);
        expect(buf.toString()).toEqual('abc');

        buf[2] += 10;
        expect(buf[2]).toEqual(109);
        expect(buf.toString()).toEqual('abm');
    });

    it('test that memory is copied from array-like', async () =>
    {
        const u = new Uint8Array(4);
        const b = Buffer.from(u);
        b[0]    = 1;
        b[1]    = 2;
        b[2]    = 3;
        b[3]    = 4;

        expect(u[0]).toEqual(0);
        expect(u[1]).toEqual(0);
        expect(u[2]).toEqual(0);
        expect(u[3]).toEqual(0);
    });

    it('detect utf16 surrogate pairs', async () =>
    {
        const text = '\uD83D\uDE38' + '\uD83D\uDCAD' + '\uD83D\uDC4D';
        const buf = Buffer.from(text);

        expect(text).toEqual(buf.toString());
    });

    it('alloc', async () =>
    {
        const data = Buffer.alloc(5);

        expect(data.length).toEqual(5);
    });
});


src/base64.ts
const lookup: any[] = [];
const revLookup: number[] = [];
const Arr = typeof Uint8Array !== 'undefined' ? Uint8Array : Array;
const INVALID_BASE64_RE = /[^+/0-9A-Za-z-_]/g;

const code = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
for (let i = 0, len = code.length; i < len; ++i) {
    lookup[i] = code[i];
    revLookup[code.charCodeAt(i)] = i;
}

// Support decoding URL-safe base64 strings, as Node.js does.
// See: https://en.wikipedia.org/wiki/Base64#URL_applications
revLookup['-'.charCodeAt(0)] = 62;
revLookup['_'.charCodeAt(0)] = 63;

export function getLens(b64: string): [number, number] {
    const len = b64.length;

    if (len % 4 > 0) {
        throw new Error('Invalid string. Length must be a multiple of 4');
    }

    // Trim off extra bytes after placeholder bytes are found
    // See: https://github.com/beatgammit/base64-js/issues/42
    let validLen = b64.indexOf('=');
    if (validLen === -1) validLen = len;

    const placeHoldersLen = validLen === len ? 0 : 4 - (validLen % 4);

    return [validLen, placeHoldersLen];
}

// base64 is 4/3 + up to two characters of the original data
export function byteLength(b64: string): number {
    const lens = getLens(b64);
    const validLen = lens[0];
    const placeHoldersLen = lens[1];
    return ((validLen + placeHoldersLen) * 3) / 4 - placeHoldersLen;
}

export function _byteLength(
    b64: string,
    validLen: number,
    placeHoldersLen: number,
): number {
    return ((validLen + placeHoldersLen) * 3) / 4 - placeHoldersLen;
}

export function toByteArray(b64: string): Uint8Array | Array<any> {
    let tmp;
    const lens = getLens(b64);
    const validLen = lens[0];
    const placeHoldersLen = lens[1];

    const arr = new Arr(_byteLength(b64, validLen, placeHoldersLen));

    let curByte = 0;

    // if there are placeholders, only get up to the last complete 4 chars
    const len = placeHoldersLen > 0 ? validLen - 4 : validLen;

    let i;
    for (i = 0; i < len; i += 4) {
        tmp =
            (revLookup[b64.charCodeAt(i)] << 18) |
            (revLookup[b64.charCodeAt(i + 1)] << 12) |
            (revLookup[b64.charCodeAt(i + 2)] << 6) |
            revLookup[b64.charCodeAt(i + 3)];
        arr[curByte++] = (tmp >> 16) & 0xff;
        arr[curByte++] = (tmp >> 8) & 0xff;
        arr[curByte++] = tmp & 0xff;
    }

    if (placeHoldersLen === 2) {
        tmp =
            (revLookup[b64.charCodeAt(i)] << 2) |
            (revLookup[b64.charCodeAt(i + 1)] >> 4);
        arr[curByte++] = tmp & 0xff;
    }

    if (placeHoldersLen === 1) {
        tmp =
            (revLookup[b64.charCodeAt(i)] << 10) |
            (revLookup[b64.charCodeAt(i + 1)] << 4) |
            (revLookup[b64.charCodeAt(i + 2)] >> 2);
        arr[curByte++] = (tmp >> 8) & 0xff;
        arr[curByte++] = tmp & 0xff;
    }

    return arr;
}

export function tripletToBase64(num: number): number {
    return (
        lookup[(num >> 18) & 0x3f] +
        lookup[(num >> 12) & 0x3f] +
        lookup[(num >> 6) & 0x3f] +
        lookup[num & 0x3f]
    );
}

export function encodeChunk(
    uint8: Uint8Array,
    start: number,
    end: number,
): string {
    let tmp;
    const output = [];
    for (let i = start; i < end; i += 3) {
        tmp =
            ((uint8[i] << 16) & 0xff0000) +
            ((uint8[i + 1] << 8) & 0xff00) +
            (uint8[i + 2] & 0xff);
        output.push(tripletToBase64(tmp));
    }
    return output.join('');
}

export function base64clean(str: string): string {
    // Node takes equal signs as end of the Base64 encoding
    str = str.split('=')[0];
    // Node strips out invalid characters like \n and \t from the string, base64-js does not
    str = str.trim().replace(INVALID_BASE64_RE, '');
    // Node converts strings with length < 2 to ''
    if (str.length < 2) return '';
    // Node allows for non-padded base64 strings (missing trailing ===), base64-js does not
    while (str.length % 4 !== 0) {
        str = str + '=';
    }
    return str;
}

export function base64ToBytes(str: string): Uint8Array | Array<any> {
    return toByteArray(base64clean(str));
}

export function fromByteArray(uint8: Uint8Array): string {
    let tmp;
    const len = uint8.length;
    const extraBytes = len % 3; // if we have 1 byte left, pad 2 bytes
    const parts = [];
    const maxChunkLength = 16383; // must be multiple of 3

    // go through the array every three bytes, we'll deal with trailing stuff later
    for (let i = 0, len2 = len - extraBytes; i < len2; i += maxChunkLength) {
        parts.push(
            encodeChunk(
                uint8,
                i,
                i + maxChunkLength > len2 ? len2 : i + maxChunkLength,
            ),
        );
    }

    // pad the end with zeros, but make sure to not forget the extra bytes
    if (extraBytes === 1) {
        tmp = uint8[len - 1];
        parts.push(lookup[tmp >> 2] + lookup[(tmp << 4) & 0x3f] + '==');
    } else if (extraBytes === 2) {
        tmp = (uint8[len - 2] << 8) + uint8[len - 1];
        parts.push(
            lookup[tmp >> 10] +
                lookup[(tmp >> 4) & 0x3f] +
                lookup[(tmp << 2) & 0x3f] +
                '=',
        );
    }

    return parts.join('');
}


src/safe-buffer.ts
// This is Buffer implementation used in SEA. Functionality is mostly
// compatible with NodeJS 'safe-buffer' and is used for encoding conversions
// between binary and 'hex' | 'utf8' | 'base64'
// See documentation and validation for safe implementation in:
// https://github.com/feross/safe-buffer#update
import BufferArray from './buffer-array';
import { base64 } from './base64';

function SafeBuffer(...props: any[])
{
    console.warn('new SafeBuffer() is depreciated, please use SafeBuffer.from()');
    return (<any>SafeBuffer).from(...props)
}

SafeBuffer.prototype = Object.create(Array.prototype);
Object.assign(SafeBuffer, {
    // (data, enc) where typeof data === 'string' then enc === 'utf8'|'hex'|'base64'
    from()
    {
        if (!Object.keys(arguments).length)
        {
            throw new TypeError(
                'First argument must be a string, Buffer, ArrayBuffer, Array, or array-like object.'
            )
        }
        const input = arguments[0];
        let buf;
        if (typeof input === 'string')
        {
            const enc = arguments[1] || 'utf8';
            if (enc === 'hex')
            {
                const bytes = (<any>input)
                    .match(/([\da-fA-F]{2})/g)
                    .map((byte: string) => parseInt(byte, 16));
                if (!bytes || !bytes.length)
                {
                    throw new TypeError('Invalid first argument for type \'hex\'.')
                }
                buf = BufferArray.from(bytes)
            }
            else if (enc === 'utf8')
            {
                const length = input.length;
                const words  = new Uint16Array(length);
                for (let i = 0; i < length; i++)
                {
                    words[i] = input.charCodeAt(i)
                }
                buf = BufferArray.from(words)
            }
            else if (enc === 'base64')
            {
                const dec    = base64.atob(input);
                const length = dec.length;
                const bytes  = new Uint8Array(length);
                for (let i = 0; i < length; i++)
                {
                    bytes[i] = dec.charCodeAt(i)
                }
                buf = BufferArray.from(bytes)
            }
            else if (enc === 'binary')
            {
                buf = BufferArray.from(input)
            }
            else
            {
                console.info('SafeBuffer.from unknown encoding: ' + enc)
            }
            return buf
        }
        const length = input?.byteLength
            ? input.byteLength
            : input?.length
                ? input?.length
                : null;

        if (length)
        {
            let buf;
            if (input instanceof ArrayBuffer)
            {
                buf = new Uint8Array(input)
            }
            return BufferArray.from(buf || input)
        }
    },
    // This is 'safe-buffer.alloc' sans encoding support
    alloc(length: number, fill = 0 /*, enc */)
    {
        return BufferArray.from(
            new Uint8Array(Array.from({ length: length }, () => fill))
        )
    },
    // This is normal UNSAFE 'buffer.alloc' or 'new Buffer(length)' - don't use!
    allocUnsafe(length: number)
    {
        return BufferArray.from(new Uint8Array(Array.from({ length: length })))
    },
    // This puts together array of array like members
    concat(arr: any[])
    {
        // octet array
        if (!Array.isArray(arr))
        {
            throw new TypeError(
                'First argument must be Array containing ArrayBuffer or Uint8Array instances.'
            )
        }
        return BufferArray.from(
            arr.reduce((ret, item) => ret.concat(Array.from(item)), [])
        )
    }
});
SafeBuffer.prototype.from     = (<any>SafeBuffer).from;
SafeBuffer.prototype.toString = BufferArray.prototype.toString;

export default <any>SafeBuffer;



src/buffer-array.ts
import { base64 } from './base64'

// This is Array extended to have .toString(['utf8'|'hex'|'base64'])
function BufferArray()
{
}

Object.assign(BufferArray, { from: Array.from });
BufferArray.prototype          = Object.create(Array.prototype);
BufferArray.prototype.toString = function (enc: string, start: number, end: number)
{
    enc          = enc || 'utf8';
    start        = start || 0;
    const length = this.length;
    if (enc === 'hex')
    {
        const buf = new Uint8Array(this);
        const num = ((end && end + 1) || length) - start;
        let res   = '';
        for (let i = 0; i < num; i++)
        {
            res += buf[i + start].toString(16).padStart(2, '0');
        }
        return res
    }
    if (enc === 'utf8')
    {
        const num = (end || length) - start;
        let res   = '';
        for (let i = 0; i < num; i++)
        {
            res += String.fromCharCode(this[i + start])
        }
        return res
    }
    if (enc === 'base64')
    {
        return base64.btoa(this)
    }
};

export default <any>BufferArray


src/utils.ts
import { base64ToBytes, fromByteArray } from './base64';
import { BaseBuffer } from './buffer';

const K_MAX_LENGTH = 0x7fffffff;

export function asciiSlice(buf: BaseBuffer, start: number, end: number): string
{
    let ret = '';
    end     = Math.min(buf.length, end);

    for (let i = start; i < end; ++i)
    {
        ret += String.fromCharCode(buf[i] & 0x7f);
    }
    return ret;
}

export function latin1Slice(buf: BaseBuffer, start: number, end: number): string
{
    let ret = '';
    end     = Math.min(buf.length, end);

    for (let i = start; i < end; ++i)
    {
        ret += String.fromCharCode(buf[i]);
    }
    return ret;
}

export function base64Slice(buf: BaseBuffer, start: number, end: number): string
{
    if (start === 0 && end === buf.length)
    {
        return fromByteArray(buf);
    }
    else
    {
        return fromByteArray(buf.slice(start, end));
    }
}

export function utf16leSlice(
    buf: BaseBuffer,
    start: number,
    end: number,
): string
{
    const bytes = buf.slice(start, end);
    let res     = '';
    // If bytes.length is odd, the last 8 bits must be ignored (same as node.js)
    for (let i = 0; i < bytes.length - 1; i += 2)
    {
        res += String.fromCharCode(bytes[i] + bytes[i + 1] * 256);
    }
    return res;
}

export function base64Write(
    buf: BaseBuffer,
    string: string,
    offset?: number,
    length?: number,
): number
{
    return blitBuffer(base64ToBytes(string), buf, offset, length);
}

export function asciiToBytes(str: string): number[]
{
    const byteArray = [];
    for (let i = 0; i < str.length; ++i)
    {
        // Node's code seems to be doing this and not & 0x7F..
        byteArray.push(str.charCodeAt(i) & 0xff);
    }
    return byteArray;
}

export function asciiWrite(
    buf: BaseBuffer,
    string: string,
    offset?: number,
    length?: number,
): number
{
    return blitBuffer(asciiToBytes(string), buf, offset, length);
}

export function utf8Write(
    buf: BaseBuffer,
    string: string,
    offset?: number,
    length?: number,
): number
{
    return blitBuffer(
        utf8ToBytes(string, buf.length - offset),
        buf,
        offset,
        length,
    );
}

export function blitBuffer(
    src: ArrayLike<any>,
    dst: BaseBuffer,
    offset?: number,
    length?: number,
): number
{
    let i;
    for (i = 0; i < length; ++i)
    {
        if (i + offset >= dst.length || i >= src.length) break;
        dst[i + offset] = src[i];
    }
    return i;
}

export function ucs2Write(buf, string, offset, length?: number)
{
    return blitBuffer(
        utf16leToBytes(string, buf.length - offset),
        buf,
        offset,
        length,
    );
}

export function utf16leToBytes(str: string, units: number): number[]
{
    let c, hi, lo;
    const byteArray: number[] = [];
    for (let i = 0; i < str.length; ++i)
    {
        if ((units -= 2) < 0) break;

        c  = str.charCodeAt(i);
        hi = c >> 8;
        lo = c % 256;
        byteArray.push(lo);
        byteArray.push(hi);
    }

    return byteArray;
}

export function hexWrite(
    buf: any,
    string: string,
    offset?: number,
    length?: number,
): number
{
    offset          = Number(offset) || 0;
    const remaining = buf.length - offset;
    if (!length)
    {
        length = remaining;
    }
    else
    {
        length = Number(length);
        if (length > remaining)
        {
            length = remaining;
        }
    }

    const strLen = string.length;

    if (length > strLen / 2)
    {
        length = strLen / 2;
    }
    let i;
    for (i = 0; i < length; ++i)
    {
        const parsed = parseInt(string.substr(i * 2, 2), 16);
        if (numberIsNaN(parsed)) return i;
        buf[offset + i] = parsed;
    }
    return i;
}

export function assertSize(size: number): void
{
    if (typeof size !== 'number')
    {
        throw new TypeError('"size" argument must be of type number');
    }
    else if (size < 0)
    {
        throw new RangeError(
            'The value "' + size + '" is invalid for option "size"',
        );
    }
}

export function fromObject(obj: any): BaseBuffer
{
    if (BaseBuffer.isBuffer(obj))
    {
        const len = checked(obj.length) | 0;
        const buf = createBuffer(len);

        if (buf.length === 0)
        {
            return buf;
        }

        obj.copy(buf, 0, 0, len);
        return buf;
    }

    if (obj.length !== undefined)
    {
        if (typeof obj.length !== 'number' || numberIsNaN(obj.length))
        {
            return createBuffer(0);
        }
        return fromArrayLike(obj);
    }

    if (obj.type === 'Buffer' && Array.isArray(obj.data))
    {
        return fromArrayLike(obj.data);
    }
}

export function fromString(string: string, encoding: string): BaseBuffer
{
    if (typeof encoding !== 'string' || encoding === '')
    {
        encoding = 'utf8';
    }

    if (!BaseBuffer.isEncoding(encoding))
    {
        throw new TypeError('Unknown encoding: ' + encoding);
    }

    const length = byteLength(string, encoding) | 0;
    let buf      = createBuffer(length);

    const actual = buf.write(string, encoding);

    if (actual !== length)
    {
        // Writing a hex string, for example, that contains invalid characters will
        // cause everything after the first invalid character to be ignored. (e.g.
        // 'abxxcd' will be treated as 'ab')
        buf = buf.slice(0, actual);
    }

    return buf;
}

export function byteLength(
    string: string|ArrayBuffer,
    encoding: string,
): number
{
    if (BaseBuffer.isBuffer(string))
    {
        return string.length;
    }
    if (ArrayBuffer.isView(string) || isInstance(string, ArrayBuffer))
    {
        return (string as ArrayBuffer).byteLength;
    }
    if (typeof string !== 'string')
    {
        throw new TypeError(
            'The "string" argument must be one of type string, Buffer, or ArrayBuffer. ' +
            'Received type ' +
            typeof string,
        );
    }

    const len       = string.length;
    const mustMatch = arguments.length > 2 && arguments[2] === true;
    if (!mustMatch && len === 0) return 0;

    // Use a for loop to avoid recursion
    let loweredCase = false;

    switch (encoding)
    {
        case 'ascii':
        case 'latin1':
        case 'binary':
            return len;
        case 'utf8':
        case 'utf-8':
            return utf8ToBytes(string).length;
        case 'ucs2':
        case 'ucs-2':
        case 'utf16le':
        case 'utf-16le':
            return len * 2;
        case 'hex':
            return len >>> 1;
        case 'base64':
            return base64ToBytes(string).length;
        default:
            if (loweredCase)
            {
                return mustMatch ? -1 : utf8ToBytes(string).length; // assume utf8
            }
            encoding    = ('' + encoding).toLowerCase();
            loweredCase = true;
    }
}

export function fromArrayLike(array: ArrayLike<any>): BaseBuffer
{
    const length = array.length < 0 ? 0 : checked(array.length) | 0;
    const buf    = createBuffer(length);
    for (let i = 0; i < length; i += 1)
    {
        buf[i] = array[i] & 255;
    }
    return buf;
}

export function checked(length: number): number
{
    // Note: cannot use `length < K_MAX_LENGTH` here because that fails when
    // length is NaN (which is otherwise coerced to zero.)
    if (length >= K_MAX_LENGTH)
    {
        throw new RangeError(
            'Attempt to allocate Buffer larger than maximum ' +
            'size: 0x' +
            K_MAX_LENGTH.toString(16) +
            ' bytes',
        );
    }
    return length | 0;
}

export function fromArrayView(arrayView: ArrayLike<any>): BaseBuffer
{
    if (isInstance(arrayView, Uint8Array))
    {
        const copy = new Uint8Array(arrayView);
        return fromArrayBuffer(copy.buffer, copy.byteOffset, copy.byteLength);
    }
    return fromArrayLike(arrayView);
}

export function fromArrayBuffer(
    array: ArrayBuffer,
    byteOffset: number,
    length: number,
): BaseBuffer
{
    if (byteOffset < 0 || array.byteLength < byteOffset)
    {
        throw new RangeError('"offset" is outside of buffer bounds');
    }

    if (array.byteLength < byteOffset + (length || 0))
    {
        throw new RangeError('"length" is outside of buffer bounds');
    }

    let buf: BaseBuffer;
    if (byteOffset === undefined && length === undefined)
    {
        buf = new BaseBuffer(array);
    }
    else if (length === undefined)
    {
        buf = new BaseBuffer(array, byteOffset);
    }
    else
    {
        buf = new BaseBuffer(array, byteOffset, length);
    }

    return buf;
}

export function utf8ToBytes(string: string, units?: number): number[]
{
    units                 = units || Infinity;
    let codePoint;
    const length          = string.length;
    let leadSurrogate     = null;
    const bytes: number[] = [];

    for (let i = 0; i < length; ++i)
    {
        codePoint = string.charCodeAt(i);

        // is surrogate component
        if (codePoint > 0xd7ff && codePoint < 0xe000)
        {
            // last char was a lead
            if (!leadSurrogate)
            {
                // no lead yet
                if (codePoint > 0xdbff)
                {
                    // unexpected trail
                    if ((units -= 3) > -1) bytes.push(0xef, 0xbf, 0xbd);
                    continue;
                }
                else if (i + 1 === length)
                {
                    // unpaired lead
                    if ((units -= 3) > -1) bytes.push(0xef, 0xbf, 0xbd);
                    continue;
                }

                // valid lead
                leadSurrogate = codePoint;

                continue;
            }

            // 2 leads in a row
            if (codePoint < 0xdc00)
            {
                if ((units -= 3) > -1) bytes.push(0xef, 0xbf, 0xbd);
                leadSurrogate = codePoint;
                continue;
            }

            // valid surrogate pair
            codePoint =
                (((leadSurrogate - 0xd800) << 10) | (codePoint - 0xdc00)) +
                0x10000;
        }
        else if (leadSurrogate)
        {
            // valid bmp char, but last char was a lead
            if ((units -= 3) > -1) bytes.push(0xef, 0xbf, 0xbd);
        }

        leadSurrogate = null;

        // encode utf8
        if (codePoint < 0x80)
        {
            if ((units -= 1) < 0) break;
            bytes.push(codePoint);
        }
        else if (codePoint < 0x800)
        {
            if ((units -= 2) < 0) break;
            bytes.push((codePoint >> 0x6) | 0xc0, (codePoint & 0x3f) | 0x80);
        }
        else if (codePoint < 0x10000)
        {
            if ((units -= 3) < 0) break;
            bytes.push(
                (codePoint >> 0xc) | 0xe0,
                ((codePoint >> 0x6) & 0x3f) | 0x80,
                (codePoint & 0x3f) | 0x80,
            );
        }
        else if (codePoint < 0x110000)
        {
            if ((units -= 4) < 0) break;
            bytes.push(
                (codePoint >> 0x12) | 0xf0,
                ((codePoint >> 0xc) & 0x3f) | 0x80,
                ((codePoint >> 0x6) & 0x3f) | 0x80,
                (codePoint & 0x3f) | 0x80,
            );
        }
        else
        {
            throw new Error('Invalid code point');
        }
    }

    return bytes;
}

export function createBuffer(length: number): BaseBuffer
{
    if (length > K_MAX_LENGTH)
    {
        throw new RangeError(
            'The value "' + length + '" is invalid for option "size"',
        );
    }
    return new BaseBuffer(length);
}

// ArrayBuffer or Uint8Array objects from other contexts (i.e. iframes) do not pass
// the `instanceof` check but they should be treated as of that type.
// See: https://github.com/feross/buffer/issues/166
export function isInstance(obj: any, type: any): boolean
{
    return (
        obj instanceof type ||
        (obj != null &&
            obj.constructor != null &&
            obj.constructor.name != null &&
            obj.constructor.name === type.name)
    );
}

export function numberIsNaN(obj: any): boolean
{
    // For IE11 support
    return obj !== obj; // eslint-disable-line no-self-compare
}

export function slowToString(encoding: string, start: number, end: number): string
{
    let loweredCase = false;

    // No need to verify that "this.length <= MAX_UINT32" since it's a read-only
    // property of a typed array.

    // This behaves neither like String nor Uint8Array in that we set start/end
    // to their upper/lower bounds if the value passed is out of range.
    // undefined is handled specially as per ECMA-262 6th Edition,
    // Section 13.3.3.7 Runtime Semantics: KeyedBindingInitialization.
    if (start === undefined || start < 0)
    {
        start = 0;
    }
    // Return early if start > this.length. Done here to prevent potential uint32
    // coercion fail below.
    if (start > this.length)
    {
        return '';
    }

    if (end === undefined || end > this.length)
    {
        end = this.length;
    }

    if (end <= 0)
    {
        return '';
    }

    // Force coercion to uint32. This will also coerce falsey/NaN values to 0.
    end >>>= 0;
    start >>>= 0;

    if (end <= start)
    {
        return '';
    }

    if (!encoding) encoding = 'utf8';

    while (true)
    {
        switch (encoding)
        {
            case 'hex':
                return hexSlice(this, start, end);

            case 'utf8':
            case 'utf-8':
                return utf8Slice(this, start, end);

            case 'ascii':
                return asciiSlice(this, start, end);

            case 'latin1':
            case 'binary':
                return latin1Slice(this, start, end);

            case 'base64':
                return base64Slice(this, start, end);

            case 'ucs2':
            case 'ucs-2':
            case 'utf16le':
            case 'utf-16le':
                return utf16leSlice(this, start, end);

            default:
                if (loweredCase)
                {
                    throw new TypeError('Unknown encoding: ' + encoding);
                }
                encoding    = (encoding + '').toLowerCase();
                loweredCase = true;
        }
    }
}

// Create lookup table for `toString('hex')`
// See: https://github.com/feross/buffer/issues/219
const hexSliceLookupTable = (function ()
{
    const alphabet = '0123456789abcdef';
    const table    = new Array(256);
    for (let i = 0; i < 16; ++i)
    {
        const i16 = i * 16;
        for (let j = 0; j < 16; ++j)
        {
            table[i16 + j] = alphabet[i] + alphabet[j];
        }
    }
    return table;
})();

function hexSlice(buf: BaseBuffer, start: number, end: number): string
{
    const len = buf.length;

    if (!start || start < 0) start = 0;
    if (!end || end < 0 || end > len) end = len;

    let out = '';
    for (let i = start; i < end; ++i)
    {
        out += hexSliceLookupTable[buf[i]];
    }
    return out;
}

export function utf8Slice(buf: BaseBuffer, start: number, end: number): string
{
    end                 = Math.min(buf.length, end);
    const res: number[] = [];

    let i = start;
    while (i < end)
    {
        const firstByte      = buf[i];
        let codePoint        = null;
        let bytesPerSequence =
                firstByte > 0xef
                    ? 4
                    : firstByte > 0xdf
                    ? 3
                    : firstByte > 0xbf
                        ? 2
                        : 1;

        if (i + bytesPerSequence <= end)
        {
            let secondByte, thirdByte, fourthByte, tempCodePoint;

            switch (bytesPerSequence)
            {
                case 1:
                    if (firstByte < 0x80)
                    {
                        codePoint = firstByte;
                    }
                    break;
                case 2:
                    secondByte = buf[i + 1];
                    if ((secondByte & 0xc0) === 0x80)
                    {
                        tempCodePoint =
                            ((firstByte & 0x1f) << 0x6) | (secondByte & 0x3f);
                        if (tempCodePoint > 0x7f)
                        {
                            codePoint = tempCodePoint;
                        }
                    }
                    break;
                case 3:
                    secondByte = buf[i + 1];
                    thirdByte  = buf[i + 2];
                    if (
                        (secondByte & 0xc0) === 0x80 &&
                        (thirdByte & 0xc0) === 0x80
                    )
                    {
                        tempCodePoint =
                            ((firstByte & 0xf) << 0xc) |
                            ((secondByte & 0x3f) << 0x6) |
                            (thirdByte & 0x3f);
                        if (
                            tempCodePoint > 0x7ff &&
                            (tempCodePoint < 0xd800 || tempCodePoint > 0xdfff)
                        )
                        {
                            codePoint = tempCodePoint;
                        }
                    }
                    break;
                case 4:
                    secondByte = buf[i + 1];
                    thirdByte  = buf[i + 2];
                    fourthByte = buf[i + 3];
                    if (
                        (secondByte & 0xc0) === 0x80 &&
                        (thirdByte & 0xc0) === 0x80 &&
                        (fourthByte & 0xc0) === 0x80
                    )
                    {
                        tempCodePoint =
                            ((firstByte & 0xf) << 0x12) |
                            ((secondByte & 0x3f) << 0xc) |
                            ((thirdByte & 0x3f) << 0x6) |
                            (fourthByte & 0x3f);
                        if (
                            tempCodePoint > 0xffff &&
                            tempCodePoint < 0x110000
                        )
                        {
                            codePoint = tempCodePoint;
                        }
                    }
            }
        }

        if (codePoint === null)
        {
            // we did not generate a valid codePoint so insert a
            // replacement char (U+FFFD) and advance only 1 byte
            codePoint        = 0xfffd;
            bytesPerSequence = 1;
        }
        else if (codePoint > 0xffff)
        {
            // encode to utf16 (surrogate pair dance)
            codePoint -= 0x10000;
            res.push(((codePoint >>> 10) & 0x3ff) | 0xd800);
            codePoint = 0xdc00 | (codePoint & 0x3ff);
        }

        res.push(codePoint);
        i += bytesPerSequence;
    }

    return decodeCodePointsArray(res);
}

// Based on http://stackoverflow.com/a/22747272/680742, the browser with
// the lowest limit is Chrome, with 0x10000 args.
// We go 1 magnitude less, for safety
const MAX_ARGUMENTS_LENGTH = 0x1000;

export function decodeCodePointsArray(codePoints: number[]): string
{
    const len = codePoints.length;
    if (len <= MAX_ARGUMENTS_LENGTH)
    {
        return String.fromCharCode.apply(String, codePoints); // avoid extra slice()
    }

    // Decode in chunks to avoid "call stack size exceeded".
    let res = '';
    let i   = 0;
    while (i < len)
    {
        res += String.fromCharCode.apply(
            String,
            codePoints.slice(i, (i += MAX_ARGUMENTS_LENGTH)),
        );
    }
    return res;
}

export function allocUnsafe(size: number): BaseBuffer
{
    assertSize(size);
    return createBuffer(size < 0 ? 0 : checked(size) | 0);
}

